### PR TITLE
[WIP] Support NHWC tensor layout

### PR DESCRIPTION
**Pull request by anaruse**
This PR aims to support **NHWC** tensor layout in convolution and batch normalization among others to improve performance when **mixed16** is set as default data type and TensorCore is used.

You can specify which parts of your model to apply **NHWC** tensor layout using ` with using_config('tensor_layout', 'NHWC') `.  The following is an example of application to RN50.

    def forward(self, x, t):
        h = self.bn1(self.conv1(x))
        h = F.max_pooling_2d(F.relu(h), 3, stride=2)
        with using_config('tensor_layout', 'NHWC'):
            h = F.transpose_to_NHWC(h)
            h = self.res2(h)
            h = self.res3(h)
            h = self.res4(h)
            h = self.res5(h)
            h = F.transpose_to_NCHW(h)
        h = F.average_pooling_2d(h, 7, stride=1)
        h = self.fc(h)
        loss = F.softmax_cross_entropy(h, t)
        chainer.report({'loss': loss, 'accuracy': F.accuracy(h, t)}, self)
        return loss

Tensor layout must be explicitly transposed to/from NHWC using ` transpose_to_NHWC/NCHW ` by users. I'm now considering how to do it automatically.

This is related to #5076.

### CODE REVIEW

**Comment by anaruse** [comment]
Please notice that this PR depends on https://github.com/cupy/cupy/pull/2235.

**Comment by anaruse** [comment]
The performance measurement results using imagenet example and resnet50 are as follows.

**[ images/sec ]**

| config | BS:32 | BS:64 | BS:128 |
|:---:|---:|---:|---:|
| float32 | 300 | 323 | 328 |
| mixed16 | 454 | 666 | 709 |
| mixed16 (nhwc) | 445 | 774 | 840 |

GPU: Quadro GV100
CUDA 10.1,  cuDNN 7.6.0,  DALI 0.10.0
nhwc layout applied to res2-5 in resnet50

**Comment by emcastillo** [comment]
NHWC support was merged in #7620
